### ISSUE TITLE
getSearchParams is not parsing correctly when using Vite

### ISSUE THREAD

**Issue by DavoCg** [kiliman/remix-params-helper#39]
Hi, 

When using Remix 2.2.0 using Vite, it seems that getSearchParams stopped parsing correctly the query params without throwing any errors.
 
It could be related to Request polyfill in vite adapater as mentionned here : https://github.com/remix-run/remix/issues/7819

**Comment by kiliman** [comment]
Thanks for the report. I'll take a look.

**Comment by kiliman** [comment]
This is weird. When I used the following code with the `remix-params-helper`, it returned an error. But when I copied the helper code directly into my project, it worked correctly.

```ts
// routes/test.$id.tsx
export async function loader({ request, params }: DataFunctionArgs) {
  const resultParams = getParams(params, z.object({ id: z.string() }))
  const resultSearchParams = getSearchParams(
    request,
    z.object({ q: z.string(), page: z.number().default(1) }),
  )
  return json({ resultParams, resultSearchParams })
}

export default function Component() {
  const loaderData = useLoaderData<typeof loader>()

  return (
    <div>
      <Form>
        <input name="q" className="border" />
      </Form>
      <pre>{JSON.stringify(loaderData, null, 2)}</pre>
    </div>
  )
}
```
### `remix-params-helper` package
<img width="473" alt="image" src="https://github.com/kiliman/remix-params-helper/assets/47168/b69b5f40-d33e-49c2-8df5-691900be3f2d">


### inline
<img width="480" alt="image" src="https://github.com/kiliman/remix-params-helper/assets/47168/86af2b8c-15d1-4eff-911e-023002a4077c">

**Comment by DavoCg** [comment]
Thanks @kiliman for your quick answer. As you mentioned, I tried copying the helper directly in my codebase and everything work as expected. This is indeed pretty weird, I'll continue investigating and update here if I find something

**Comment by kiliman** [comment]
I did some more investigation. It looks like I need to make a pure ESM package. By default it was loading the CJS version, and for some reason, this code was always returning `false`, so it wasn't parsing properly and always failing.

```ts
function parseParams(o, schema, key, value) {
    var _a;
    // find actual shape definition for this key
    let shape = schema;
    console.log('parseParams', shape instanceof zod_1.ZodObject)
    while (shape instanceof zod_1.ZodObject || shape instanceof zod_1.ZodEffects) {
        shape =
            shape instanceof zod_1.ZodObject
                ? shape.shape
                : shape instanceof zod_1.ZodEffects
                    ? shape._def.schema
                    : null;
        if (shape === null) {
            throw new Error(`Could not find shape for key ${key}`);
        }
    }
    // ...
}
```

**Comment by DavoCg** [comment]
I was about to send the same finding after debugging in node_modules, ```const def = shape[key];``` is `undefined` and not processed. I'll use a local version in the meantime, thanks for taking the time @kiliman !

**Comment by mauscoelho** [comment]
Sorry for recommending something different here. I needed a quick fix and started using [Zodix](https://github.com/rileytomasek/zodix). 
The API is the same so it was an easy migration.

**Comment by kiliman** [comment]
@mauscoelho I'm glad you found something that works for you.

**Comment by hi-ogawa** [comment]
> It looks like I need to make a pure ESM package. By default it was loading the CJS version, and for some reason, this code was always returning false, so it wasn't parsing properly and always failing.

I think the issue is a dual package hazard (see https://nodejs.org/api/packages.html#dual-package-hazard) of zod dependency where user code loads esm version of zod (via `import`) but `remix-params-helper` loads cjs version (via `require`).

I was interested in this issue since this might be a general problem which others libraries might face.
For the specific case of `remix-params-helper`, it already includes esm version in the package, so enforcing vite to resolve esm version can be a quick workaround available on user side.
This can be achieved by the combination of `ssr.noExternal` and `resolve.alias` like this:
https://stackblitz.com/edit/remix-run-remix-cktyfl?file=vite.config.ts

```tsx
export default defineConfig({
  plugins: [remix(), tsconfigPaths()],
  ssr: {
    noExternal: ['remix-params-helper'],
  },
  resolve: {
    alias: {
      'remix-params-helper': 'node_modules/remix-params-helper/dist/helper.js',
    },
  },
});
```

<details><summary>reveal screenshot</summary>

![image](https://github.com/kiliman/remix-params-helper/assets/4232207/8ccabe03-ca7b-4a56-8448-4ad8c7e03156)

</details>

I'm not familiar with `remix-params-helper` myself so I might missing some scenarios not covered by this workaround, but I hope this still gives some ideas of the underlying issue!

**Comment by Skoob1905** [comment]
+1 to this. an error is always returned. I'm trying to use this in my loader to parse queryParams for my API but just throws an error everytime.

**Comment by kiliman** [comment]
I'll be publishing an updated package that works properly for both CJS and ESM imports.

**Comment by kiliman** [comment]
Fixed in v0.5.0
